ndb_api/db_helpers.go

```diff
@@ -304,6 +305,202 @@ func (a *PostgresRequestAppender) appendProvisioningRequest(req *DatabaseProvisi
+			NetworkProfileId: req.NetworkProfileId,
+			ComputeProfileId: req.ComputeProfileId,
+		})
+	}
+	const MinReqDatabaseNodes = 3
+	if nodeCount < MinReqDatabaseNodes {
+		return fmt.Errorf("invalid node count: HA instance needs at least %d nodes, given: %d", MinReqDatabaseNodes, nodeCount)
+	}
+
+	return nil
+}
+
+func createDefaultNodes(database DatabaseInterface) []*v1alpha1.Node {
+	nodes := make([]*v1alpha1.Node, 0)
+	nodes = append(nodes, &v1alpha1.Node{
+		VmName: database.GetAdditionalArguments()["cluster_name"] + "_haproxy1",
+		Properties: v1alpha1.NodeProperties{
+			NodeType: "haproxy",
+		},
+	})
+	nodes = append(nodes, &v1alpha1.Node{
+		VmName: database.GetAdditionalArguments()["cluster_name"] + "_haproxy2",
+		Properties: v1alpha1.NodeProperties{
+			NodeType: "haproxy",
+		},
+	})
+	nodes = append(nodes, &v1alpha1.Node{
+		VmName: database.GetAdditionalArguments()["cluster_name"] + "-1",
+		Properties: v1alpha1.NodeProperties{
+			NodeType:     "database",
+			Role:         "Primary",
+			FailoverMode: "Automatic",
+		},
+	})
+	nodes = append(nodes, &v1alpha1.Node{
+		VmName: database.GetAdditionalArguments()["cluster_name"] + "-2",
+		Properties: v1alpha1.NodeProperties{
+			NodeType:     "database",
+			Role:         "Secondary",
+			FailoverMode: "Automatic",
+		},
+	})
+	nodes = append(nodes, &v1alpha1.Node{
+		VmName: database.GetAdditionalArguments()["cluster_name"] + "-3",
+		Properties: v1alpha1.NodeProperties{
+			NodeType:     "database",
+			Role:         "Secondary",
+			FailoverMode: "Automatic",
+		},
+	})
+	return nodes
+}
+
+func getPrimaryNodeCount(nodesRequested []*v1alpha1.Node) int {
+	count := 0
+	for _, node := range nodesRequested {
+		if node.Properties.Role == "Primary" {
+			count++
+		}
+	}
+	return count
+}
+
+func (a *PostgresHARequestAppender) appendProvisioningRequest(req *DatabaseProvisionRequest, database DatabaseInterface, reqData map[string]interface{}) (*DatabaseProvisionRequest, error) {
+	dbPassword := reqData[common.NDB_PARAM_PASSWORD].(string)
+	databaseNames := database.GetInstanceDatabaseNames()
+	req.SSHPublicKey = reqData[common.NDB_PARAM_SSH_PUBLIC_KEY].(string)
+	// Set the number of nodes to 5, 3 Postgres nodes + 2 HA Proxy nodes
+	err := setNodesParameters(req, database)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Clustered = true
+
+	// Default action arguments
+	actionArguments := defaultActionArgumentsforHAProvisioning(database, dbPassword, databaseNames)
+
+	// Appending/overwriting database actionArguments to actionArguments
+	if err := setConfiguredActionArguments(database, actionArguments); err != nil {
+		return nil, err
+	}
+	// Converting action arguments map to list and appending to req.ActionArguments
+	req.ActionArguments = append(req.ActionArguments, convertMapToActionArguments(actionArguments)...)
+
+	return req, nil
+}
+
+func defaultActionArgumentsforHAProvisioning(database DatabaseInterface, dbPassword string, databaseNames string) map[string]string {
```

Comment:
move the above function and remain consistent with the rest of the appenders, as indicated in the top comment.